In [39]:
# 파일 불러오기

import numpy as np
import pandas as pd

score = pd.read_csv('./dataset/score.csv')
user = pd.read_csv('./dataset/user.csv')


In [13]:
print(score.info())
print(score.shape)
score.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   sno     14 non-null     int64  
 1   kor     12 non-null     float64
 2   eng     11 non-null     float64
 3   math    14 non-null     int64  
 4   year    14 non-null     int64  
dtypes: float64(2), int64(3)
memory usage: 688.0 bytes
None
(14, 5)


sno   kor   eng  math  year
0   1001  88.0  84.0    92  2020
1   1001   NaN  79.0    82  2021
2   1002  84.0  88.0    94  2020
3   1002  74.0   NaN    79  2021
4   1003  92.0  94.0    90  2020
5   1003  86.0   NaN    84  2021
6   1004  88.0  84.0    88  2020
7   1004  79.0  82.0    74  2021
8   1005  90.0  94.0    90  2020
9   1005  84.0   NaN    86  2021
10  1006  88.0  90.0    92  2020
11  1006  74.0  82.0    79  2021
12  1007  92.0  84.0    94  2020
13  1007   NaN  84.0    74  2021

In [14]:
print(user.info())
print(user.shape)
user.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   sno     7 non-null      int64 
 1   ename   7 non-null      object
 2   age     7 non-null      int64 
 3   gen     7 non-null      object
dtypes: int64(2), object(2)
memory usage: 352.0+ bytes
None
(7, 4)


sno ename  age gen
0  1001   송강호   33   m
1  1002   황정민   29   m
2  1003   김윤석   31   m
3  1004   설경구   27   m
4  1005   김혜수   32   f

데이터 읽으면 데이터 관계 파악★ -> sno로 이을 수 있겠다. score는 학생 당 2개년치 있어서 14행, 학생은 7행

### 1. 각 과목별 결측 데이터 수

In [25]:
# score.columns
# score.isna().sum() # --> 이게 정석(추천)
score[['kor', 'eng', 'math']].isna().sum()

kor     2
eng     3
math    0
dtype: int64

### 2. 2020년도 각 과목별 평균 점수

In [33]:
# score[컬럼][조건]
# score[조건][컬럼]

score[score['year']==2020].groupby('year')[['kor', 'eng', 'math']].mean()
## 가운데 .groupby 안 해도 되네

kor        eng       math
year                                 
2020  88.857143  88.285714  91.428571

* 3. 성적이 없는 학생의 점수는 '50점'               로 채우기
* 3. 성적이 없는 학생의 점수는 '해당 과목의 최저 점수'로 채우기

In [38]:
score50=score.fillna(50).copy()
score50

sno   kor   eng  math  year
0   1001  88.0  84.0    92  2020
1   1001  50.0  79.0    82  2021
2   1002  84.0  88.0    94  2020
3   1002  74.0  50.0    79  2021
4   1003  92.0  94.0    90  2020
5   1003  86.0  50.0    84  2021
6   1004  88.0  84.0    88  2020
7   1004  79.0  82.0    74  2021
8   1005  90.0  94.0    90  2020
9   1005  84.0  50.0    86  2021
10  1006  88.0  90.0    92  2020
11  1006  74.0  82.0    79  2021
12  1007  92.0  84.0    94  2020
13  1007  50.0  84.0    74  2021

In [55]:
# 틀렸다가 다시 푸니까 맞았는데 score.min() 완벽 이해는 못 함. -> 밑에 혼자 공부 참고

smin=score[['kor', 'eng', 'math']].min().tolist()
print(smin)
score_min=score.fillna(score.min()).copy()
score_min


[74.0, 79.0, 74.0]


sno   kor   eng  math  year
0   1001  88.0  84.0    92  2020
1   1001  74.0  79.0    82  2021
2   1002  84.0  88.0    94  2020
3   1002  74.0  79.0    79  2021
4   1003  92.0  94.0    90  2020
5   1003  86.0  79.0    84  2021
6   1004  88.0  84.0    88  2020
7   1004  79.0  82.0    74  2021
8   1005  90.0  94.0    90  2020
9   1005  84.0  79.0    86  2021
10  1006  88.0  90.0    92  2020
11  1006  74.0  82.0    79  2021
12  1007  92.0  84.0    94  2020
13  1007  74.0  84.0    74  2021

In [91]:
##################
# 혼자 공부한 것 #
##################

# 결과값 저장은 생략

# ------ score 복제 df 만들기

scoretttt = pd.read_csv('./dataset/score.csv')
scoretttt.isna().sum()


# ------ key를 column 이름으로 하는 임시 딕셔너리 만들기(value는 na일 경우 채울 값)

tempd={'kor':9999}


# ------ fillna에 딕셔너리를 넣으면 각 column의 이름과 같은 이름의 'key'가 연결되어 해당 column의 na에 value가 채워진다.

scoretttt.fillna(tempd)


# ------ index를 column 이름으로 하는 임시 시리즈 만들기(값은 na일 경우 채울 값)

temps=pd.Series([6666], index=['eng'])


# ------ fillna에 시리즈를 넣으면 각 column의 이름과 같은 이름의 index가 연결되어 해당 column의 na에 값이 채워진다.

scoretttt.fillna(temps)


# ------ -----응용!!
# ------ 그룹함수 사용하기

scoretttt.min() #그룹함수를 사용하면 결과는 series로 나온다.
scoretttt.fillna(scoretttt.min()) # Na가 각 column의 최소값으로 채워진다.

scoretttt.min().to_dict() # dict로 해도 마찬가지! (굳이..?)
scoretttt.fillna(scoretttt.min().to_dict())

sno   kor   eng  math  year
0   1001  88.0  84.0    92  2020
1   1001  74.0  79.0    82  2021
2   1002  84.0  88.0    94  2020
3   1002  74.0  79.0    79  2021
4   1003  92.0  94.0    90  2020
5   1003  86.0  79.0    84  2021
6   1004  88.0  84.0    88  2020
7   1004  79.0  82.0    74  2021
8   1005  90.0  94.0    90  2020
9   1005  84.0  79.0    86  2021
10  1006  88.0  90.0    92  2020
11  1006  74.0  82.0    79  2021
12  1007  92.0  84.0    94  2020
13  1007  74.0  84.0    74  2021

In [ ]:
# 떼거지로 했을 때 inplace=True 하면 제대로 안 먹는다!!

In [58]:
score['kor'].min()
score['eng'].min()

79.0

In [59]:
score1 = score.copy()
#score1['kor'].fillna(50, inplace=True)
#score1['eng'].fillna(50, inplace=True)
score1[  ['kor','eng'] ] = score1[  ['kor','eng'] ].fillna(50)
score1.isna().sum()


sno     0
kor     0
eng     0
math    0
year    0
dtype: int64

In [60]:
# 성적이 없는 학생의 점수는 '해당 과목의 최저 점수'로 채우기
score['kor'].fillna(score['kor'].min(), inplace=True)
score['eng'].fillna(score['eng'].min(), inplace=True)


In [61]:
score.isna().sum()


sno     0
kor     0
eng     0
math    0
year    0
dtype: int64

### 4. kor + eng 평균점수 lang 컬럼 추가

In [68]:
# score['lang']=(score['kor'] + score['eng'])/2
# score.head()

# 보너스로 타입 바꾸기
score[['kor','eng', 'lang']]=score[['kor','eng','lang']].astype('int')
score.head()

sno  kor  eng  math  year  lang
0  1001   88   84    92  2020    86
1  1001   74   79    82  2021    76
2  1002   84   88    94  2020    86
3  1002   74   79    79  2021    76
4  1003   92   94    90  2020    93

### 5. 모든 수강생의 국,영,수 점수 (년도,학번 정렬)

In [71]:
score.sort_values(by=['year', 'sno'])

sno  kor  eng  math  year  lang
0   1001   88   84    92  2020    86
2   1002   84   88    94  2020    86
4   1003   92   94    90  2020    93
6   1004   88   84    88  2020    86
8   1005   90   94    90  2020    92
10  1006   88   90    92  2020    89
12  1007   92   84    94  2020    88
1   1001   74   79    82  2021    76
3   1002   74   79    79  2021    76
5   1003   86   79    84  2021    82
7   1004   79   82    74  2021    80
9   1005   84   79    86  2021    81
11  1006   74   82    79  2021    78
13  1007   74   84    74  2021    79

<pre>
sno ename  age gen  kor  eng  math  year  lang
0   1001   송강호   33   m   88   84    92  2020    86
2   1002   황정민   29   m   84   88    94  2020    86
4   1003   김윤석   31   m   92   94    90  2020    93
6   1004   설경구   27   m   88   84    88  2020    86
8   1005   김혜수   32   f   90   94    90  2020    92
10  1006   하정우   23   m   88   90    92  2020    89
12  1007   전지현   26   f   92   84    94  2020    88
1   1001   송강호   33   m   74   79    82  2021    76
3   1002   황정민   29   m   74   79    79  2021    76
5   1003   김윤석   31   m   86   79    84  2021    82
7   1004   설경구   27   m   79   82    74  2021    80
9   1005   김혜수   32   f   84   79    86  2021    81
11  1006   하정우   23   m   74   82    79  2021    78
13  1007   전지현   26   f   74   84    74  2021    79


eng    6666
dtype: int64